implementations/rust/ockam/ockam_api/src/kafka/kafka_portal_worker.rs

```diff
@@ -0,0 +1,423 @@
+
+enum Receiving {
+    Requests,
+    Responses,
+}
+
+///Acts like a relay for messages between tcp inlet and outlet for both directions.
+/// It's meant to be created by the portal listener.
+///
+/// This instance manage both streams inlet and outlet in two different workers, one dedicated
+/// to the requests (inlet=>outlet) the other for the responses (outlet=>inlet).
+/// since every kafka message is length-delimited every message is read and written
+/// through a framed encoder/decoder.
+///
+/// ```text
+/// ┌────────┐  decoder    ┌─────────┐  encoder    ┌────────┐
+/// │        ├────────────►│ Kafka   ├────────────►│        │
+/// │        │             │ Request │             │        │
+/// │  TCP   │             └─────────┘             │  TCP   │
+/// │ Inlet  │             ┌─────────┐             │ Outlet │
+/// │        │  encoder    │ Kafka   │   decoder   │        │
+/// │        │◄────────────┤ Response│◄────────────┤        │
+/// └────────┘             └─────────┘             └────────┘
+///```
+pub(crate) struct KafkaPortalWorker {
+    other_worker_address: Address,
+    reader: KafkaDecoder,
+    writer: KafkaEncoder,
+    receiving: Receiving,
+    shared_protocol_state: Arc<ProtocolState>,
+    inlet_map: Arc<KafkaInletMap>,
+}
+
+#[ockam::worker]
+impl Worker for KafkaPortalWorker {
+    type Message = PortalMessage;
+    type Context = Context;
+
+    async fn handle_message(
+        &mut self,
+        context: &mut Self::Context,
+        routed_message: Routed<Self::Message>,
+    ) -> ockam::Result<()> {
+        trace!("received message: {:?}", &routed_message);
+
+        let onward_route = routed_message.onward_route();
+        let return_route = routed_message.return_route();
+        let portal_message = PortalMessage::decode(routed_message.payload())?;
+
+        match portal_message {
+            PortalMessage::Payload(message) => {
+                let result = self
+                    .intercept_and_transform_messages(context, message)
+                    .await;
+                if let Err(cause) = result {
+                    return Err(Error::new(Origin::Transport, Kind::Io, cause));
+                }
+
+                if let Some(encoded_message) = result.unwrap() {
+                    self.split_and_send(
+                        context,
+                        onward_route,
+                        return_route,
+                        encoded_message,
+                        routed_message.local_message().local_info(),
+                    )
+                    .await?;
+                }
+            }
+            PortalMessage::Disconnect => {
+                //the shutdown will take care of sending disconnect
+                context
+                    .stop_worker(self.other_worker_address.clone())
+                    .await?;
+                context.stop_worker(context.address()).await?;
+                self.forward(context, routed_message).await?;
```

Comment:
I'm not sure it's safe to use `Context` instance after it was shutdown

Reply:
make sense, reversing the order in doubt!